Merge request of large scale developments to code base

badeaa3 opened the pull request:
Changes include: 

- Cleaning up main simulation code. 
- Movement of segments of code into functions which are stored in header files. 
- Separation of code based on their role into header files. 
- Legacy support for previous versions of code.
- Large amounts of comments added. 
- Interface to write and load data and simulation argument branches for ntuple production and reading. 
- Addition of simulation arguments into ttrees. 
- Script to read stdout file and launch jobs with the parameters from it. 
- Added feature to take in 8, 64, 128 MM efficiencies for planes, PCBs, and L/R PCBs respectively. 
- Automatically validation script to compare different productions and simulations. Slides are automatically produced via this script as well. 
- Addition of several comprehensive plotting scripts to produce nice looking plots. 

Other changes are present as well but these constitute the major changes.

Commit:
parallel data generation

Commit:
batch job data generation script

badeaa3 committed:
data gneration update

badeaa3 committed:
updating batch script

badeaa3 committed:
updated data generation

badeaa3 committed:
updated generateDataBatch.py

badeaa3 committed:
first simulation studies

badeaa3 committed:
Uploading Andreas' old analysis code for savekeeping

badeaa3 committed:
Updated plotting macros

badeaa3 committed:
Fixed plotting bug

badeaa3 committed:
enable each mm eff to be tuned individually

badeaa3 committed:
Merge branch 'master' of https://github.com/badeaa3/oct_sim

badeaa3 committed:
Adding a comments part 1

badeaa3 committed:
Update to EfficiencyPlots.ipynb

badeaa3 committed:
Updated simulation production to allow mm plane efficiencies to vary: BUG FIX

badeaa3 committed:
Updating the github

badeaa3 committed:
Big update to produce modular simulation code

badeaa3 committed:
Updating some comments

badeaa3 committed:
Updating efficiency input from user

badeaa3 committed:
Modification of batch script for production on local computers and script to batch produce with different job arguments

badeaa3 committed:
Added TTree sim_args containing the simulation arguments

badeaa3 committed:
Moved tree declaration, filling, and loading into a class in header file SimNtupleData.hh

badeaa3 committed:
Update to ntuple and batch production

badeaa3 committed:
New python script to read stdout files and reproduce the run

badeaa3 committed:
input argument parser

badeaa3 committed:
Update to stdout writing

badeaa3 committed:
Adding some comments

badeaa3 committed:
Added in more legacy support options

badeaa3 committed:
Make kill_random output an std::vector<int>, before it was operating on an input vector rather than a pointer to the original vector. Better to fix this with an explicit return of the vector rather than operate on the old vector so that we have access to the old vector in the main simulation script for debugging purposes

badeaa3 committed:
Update validation plotting

badeaa3 committed:
remove old file

badeaa3 committed:
Big plotting update for validation

badeaa3 committed:
generateData.py

sezata commented:
I'm happy to merge - this is an enormous amount of work! If you have time, could you also update the README? There's a ton of changes and different analysis scripts now. I can also help out if needed.

annmwang commented on include/styleUtil.h:
```diff
@@ -0,0 +1,847 @@
+
+void setCanvas_InvMass_pp(TCanvas* c, float px, float py, const char* ptCut, const char* etaCut, const char* extra)
+{
+    setCanvas_InvMass(c, px, py, "CMS preliminary pp", "#sqrt{s} = 5.02 TeV", ptCut, etaCut, extra);
```
This script seems to have some extraneous CMS stuff. Are you using this file anywhere?

badeaa3 replied:
I use it for some of the histogram formatting. I'm deleting all the CMS specific code.

annmwang commented on python/batch_local.py:
```diff
@@ -0,0 +1,162 @@
+"""
+batch.py :: a script for launching sim jobs
```
What's the difference between this and batch.py? Can we get rid of one?

badeaa3 replied:
batch.py has parameters specific for NET3 and whichever cluster used slurm. This was to have a script which could run locally instead of on the servers. I was planning to add that additional feature into batch.py at some point. I'll do that.

annmwang commented on src/plot_sim.C:
```diff
@@ -0,0 +1,68 @@
+      t_args->GetEntry(i);
+      //std::cout<< SN.bkgrate << std::endl;
+    }
+    std::cout<< SN.real_x_muon << std::endl;
```
Is this file doing anything? It seems like it's just reading in a ttree and printing out a number.

badeaa3 replied:
I deleted this file. It was to test if the class SimNtuple worked properly.

annmwang commented on oct_sim_analysis/modplot.py:
```diff
@@ -0,0 +1,208 @@
+    return defaults
+
+# default style for cms data
+def cms_style(*args, **kwargs):
```
Same comment for this - seems like extraneous utilities

badeaa3 replied:
You're right so I removed this entire folder. This is personal analysis stuff that I'll keep on my fork in a dev branch later.

annmwang reviewed (commented):
Added a few comments but still happy to merge. A little unclear if all of the different plotters and jupyter notebooks are necessary.

badeaa3 committed:
Updating the merge pull request based on comments

badeaa3 commented:
Just resolved the comments by @sezata in the newest push 4c4090c. When @sezata and @alexandertuna review again then we can finalize the merge.

alexandertuna commented on include/SimUtilities.hh:
```diff
@@ -0,0 +1,438 @@
+																			  TRandom3 *ran, // random number generator from main method
+																			  int sig_art, // art time resolution from set_chamber (in ns)
+																			  int m_sig_art_x, // art time resolution (in ns)
+																			  int bc_length, // time between bunch crossing (in ns)
```
Was there a sale on whitespace?

alexandertuna commented:
Hey Anthony, really nice stuff. The modularity brings me immense joy.

I'm happy to merge whenever you're ready, with one excpetion: I would greatly prefer you address my first comment (tabs and spaces). The remaining comments are smaller. (e.g. I'm greatly interested in an efficiency plot, but I don't need to see it before merging this.)

Here's a little feedback:

- General
  - Nice
  - I think this mixes tabs with spaces. Can you use spaces everywhere?
    - Most editors have a quick way to replace tabs with spaces
  - Have you made plot of efficiency, as part of the cross-check?
    - This would bring me great comfort
  - Feels like a big dump of plotting stuff
    - Which is fine, since you're the only person using the code right now, but
    - As you've seen, it's not easy for new people to learn large swaths of code
    - If you can simplify anywhere, I'm sure it will be appreciated
- Also, personally, I like separating histogram/tree-making from histogram/tree-plotting
    - Making: C++
    - Plotting: python
    - But you're in charge. I don't object to C++ plotters (like styleUtil.h)
- I strongly support moving away from using cout for anything - thank you for this
  - Put everything into root file (TTree)
  - Analyzer only needs 1 source (TTree) instead of 2 sources (TTree and cout)
  - No possibility of conflict between TTree and cout
- `include/ChamberUtilities.hh`
  - Why `struct slope_t` and `bool compare_slope` here? 
  - Not really chamber-related. I think this would fit better in `Finder.hh`
- `HttStyles`
  - lmao
- Do you have two github accounts?
  - badeaa3 vs badeaA
  - ???

badeaa3 committed:
Removing white space and changing all indentation to spaces

badeaa3 committed:
Move  and  to

badeaa3 commented:
Hi @alexandertuna, 

- I removed as much white space as I could see on a first pass and changed the indentation to spacing. 
- I have made a plot of efficiency as in Figure 4 using the old simulation code. The code for this plot is not included in the repo, however, because we didn't think for a first merge that it was core simulation code. Given the outputs of the new code and old code are the same, which they are, the plot of efficiency will be identical. I can remake it though. 
- The big dump of plotting scripts I could likely clean up a bunch but haven't gone through it and would rather first merge then clean so I don't have to touch/delete working code. Also the plotting functionality was very helpful for making the automatic slides. 
- I somewhat agree with histograms in C++ and plotting in python and have been told that by you, Ann, and Larry. The reason I didn't do that for the validation script is cause I already had code written to produce latex slides in the way I did. It was just easier for me to do this way. 
- Moving away from cout is done and when the time comes I'll make an equivalent ttree reader as I have a cout reader to launch jobs using the parameters in a tree. 
- Thanks for checking the methods in each of the header files. I moved `struct slope_t` and `bool compare_slope` to `Finder.hh`. 
- I think the two different GitHub names are just from me configuring my name differently on lxplus and locally. Not really sure which is which but I'll look into it. Probably should check that out. 

Anthony

badeaa3 commented:
@alexandertuna @sezata Following up on your comment earlier about the trigger efficiency plot. My intentions are to add into the validation plotting script any other plots that we want so that the slides will automatically include them. Then we can have a pipeline that will run the job and make slides in one command. :)

sezata commented:
🎉
